File: LeetCode/216/main.py
class Solution:
    def combinationSum3(self, k: int, n: int) -> List[List[int]]:
        
      

        def solve(ind, pref, target):
            if(target == 0 and len(pref) == k): 
                # Copy of pref
                ans.append(pref[:])    
                return 

            for i in range(ind,len(arr)):
                if target < arr[i]:
                    break

                # Set the state
                pref.append(arr[i])     
                
                # Goto next state
                solve(i+1, pref, target - arr[i])  
                
                # Reset the state
                pref.pop()       

        if k > n:   
            return []
        if n > 45:   
            return []
            
        arr = range(1,10)
        ans = []
        solve(0,[],n)
        
        return ans

File: LeetCode/216/main.cpp
class Solution {
public:
  vector<vector<int>> ans;

  void helper(vector<int> &path, int pos, int k, int target) {
    if (k == 0) {
      if (target == 0) {
        ans.emplace_back(path);
      }
      return;
    }

    for (int i = pos; i < 10; i++) {

      // set state
      path.emplace_back(i);

      // goto next state
      helper(path, i + 1, k - 1, target - i);

      // reset state
      path.pop_back();
    }

    return;
  }

  vector<vector<int>> combinationSum3(int k, int n) {
    vector<int> path;

    helper(path, 1, k, n);

    return ans;
  }
};